tower-http/src/metrics/traffic/metrics_sink.rs

```diff
@@ -0,0 +1,113 @@
+    fn prepare<B>(&mut self, request: &Request<B>) -> Self::Data;
+
+    /// Perform some action when a response has been generated.
+    ///
+    /// This method is called when the inner [`Service`]'s response future completes with
+    /// `Ok(response)`, regardless if the response is classified as a success or a failure.
+    ///
+    /// If the response is the start of a stream (as determined by the classifier passed to
+    /// [`Traffic::new`] or [`TrafficLayer::new`]) then `classification` will be
+    /// [`ClassifiedResponse::RequiresEos(())`], otherwise it will be
+    /// [`ClassifiedResponse::Ready`].
+    ///
+    /// The default implementation does nothing and returns immediately.
+    ///
+    /// [`Traffic::new`]: crate::metrics::Traffic::new
+    /// [`TrafficLayer::new`]: crate::metrics::TrafficLayer::new
+    /// [`ClassifiedResponse::RequiresEos(())`]: crate::classify::ClassifiedResponse::RequiresEos
+    /// [`Service`]: tower::Service
+    #[inline]
+    #[allow(unused_variables)]
+    fn on_response<B>(
+        &mut self,
+        response: &Response<B>,
+        classification: ClassifiedResponse<FailureClass, ()>,
+        data: &mut Self::Data,
+    ) {
+    }
+
+    /// Perform some action when a stream has ended.
+    ///
+    /// This is called when [`Body::poll_trailers`] completes with `Ok(trailers)` regardless if
+    /// the trailers are classified as a failure.
+    ///
+    /// A stream that ends successfully will trigger two callbacks. [`on_response`] will be called
+    /// once the response has been generated and the stream has started and [`on_eos`] will be
+    /// called once the stream has ended.
+    ///
+    /// If the trailers were classified as a success then `classification` will be `Ok(())`
+    /// otherwise `Err(failure_class)`.
+    ///
+    /// The default implementation does nothing and returns immediately.
+    ///
+    /// [`on_response`]: MetricsSink::on_response
+    /// [`on_eos`]: MetricsSink::on_eos
+    /// [`Body::poll_trailers`]: http_body::Body::poll_trailers
+    #[inline]
+    #[allow(unused_variables)]
+    fn on_eos(
+        self,
+        trailers: Option<&HeaderMap>,
+        classification: Result<(), FailureClass>,
+        data: Self::Data,
+    ) {
+    }
+
+    /// Perform some action when an error has been encountered.
+    ///
+    /// This method is only called in these scenarios:
+    ///
+    /// - The inner [`Service`]'s response future resolves to an error.
+    /// - [`Body::poll_data`] returns an error.
+    /// - [`Body::poll_trailers`] returns an error.
+    ///
+    /// That means this method is _not_ called if a response is classified as a failure (then
+    /// [`on_response`] is called) or an end-of-stream is classified as a failure (then [`on_eos`]
+    /// is called).
+    ///
+    /// `failed_at` specifies where the error happened.
+    ///
+    /// The default implementation does nothing and returns immediately.
+    ///
```

Comment:
I'm a bit unsure if this design is actually correct. Specifically I'm wondering if `Body::poll_data` can return multiple failures. If so that'll require changes to the classification. I've asked on Discord. Until thats been cleared we cannot merge this.